### PR TITLE
Add checking access to memory

### LINKED ISSUE #2132: WASM-GEN: check pointer & size passed to syscalls
### Problem to Solve

avoid too often OOB errors

<details><summary>Click to expand/collapse</summary>
<p>

```
2023-01-25 14:43:36.021 DEBUG tokio-runtime-worker gear_core_processor::processor: Charged for module instantiation and memory pages. Size: WasmPage(168)    
2023-01-25 14:43:36.021 DEBUG tokio-runtime-worker gear_core_processor::executor: Executing program 0x05bad3b550f45b8eb8f8acb172011581c37ca3bee0bd0888b1a47926f805c35f    
2023-01-25 14:43:36.021 DEBUG tokio-runtime-worker gear_core_processor::executor: Executing dispatch IncomingDispatch { kind: Init, message: IncomingMessage { id: 0xa663832d1b7bd2f4bdbfe82cf0877a1cc5204f7aee89b5758da917ed0981b08d, source: 0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d, payload: 0xbfc0099f8db366762a2028d0d4694ad808d8d4c7723f65dba32f09db6c5c9fc7776b0d53b156dd4836ea0317e5e1697c098f372a73a6ba191075456f21dacd2bba098910e0b65b2187f2b93099b62c201a61e268f14d092897ea4f58f54b657ad30d, gas_limit: 25000000000, value: 0, details: None }, context: None }    
2023-01-25 14:43:36.025 TRACE tokio-runtime-worker syscall::gear: debug, args =  I32(0x0), I32(0x28),    
2023-01-25 14:43:36.025 TRACE tokio-runtime-worker syscall::gear: reply_wgas, args =  I32(0x125346d), I32(0xc509defe), I64(0x27f45ba333), I32(0x1092961), I32(0x1bd), I32(0x5312d2),    
2023-01-25 14:43:36.057 DEBUG tokio-runtime-worker gear_backend_sandbox::env: SandboxEnvironment::execute res = Err(<wasm:stripped>)    
2023-01-25 14:43:36.057 DEBUG tokio-runtime-worker gear_core_processor::executor: Termination reason: Trap(Other(TrimmedString("Memory access out of bounds")))    
2023-01-25 14:43:36.057 DEBUG tokio-runtime-worker gear_core_processor::executor: 💥 Trap during execution of 0x05bad3b550f45b8eb8f8acb172011581c37ca3bee0bd0888b1a47926f805c35f
📔 Explanation: Memory access out of bounds    
2023-01-25 14:43:36.057 DEBUG tokio-runtime-worker gear_common::gas_provider::internal: Spent 16967757814 of gas    
```

```
with seed: 749505056160585358,
```
</p>
</details>

### Possible Solution

check that pointer + size <= memory

### Notes

inspired by #1564

### PR DESCRIPTION
**Pull request by gshep**
Resolves #2132 

- introduce enumeration Parameter to perform memory access checking if required
- add wasm module with page size constant and helper funcs
- doc comments & a bit of refactoring

@gear-tech/dev

### CODE REVIEW

**Review by SabaunT — APPROVED** [review]

File: utils/wasm-gen/src/syscalls.rs [hunk 1]
```diff
@@ -87,43 +130,43 @@ pub struct SyscallsConfig {
 
 impl Default for SyscallsConfig {
     fn default() -> Self {
-        let restricted_ratio = (1, 100).into();
+        let unrestricted_ratio = (1, 100).into();
```
[the remark this answers is not in the document]
Comment (gshep) [reply utils/wasm-gen/src/syscalls.rs:133]: discussed

File: utils/wasm-instrument/src/syscalls.rs [hunk 2]
```diff
@@ -193,6 +193,8 @@ impl SysCallName {
             Self::Read,
             Self::Reply,
             Self::ReplyWGas,
+            Self::ReplyInput,
```
Comment (breathx) [inline utils/wasm-instrument/src/syscalls.rs:196]: how could it be wout failing on the backend side?

Comment (breathx) [reply utils/wasm-instrument/src/syscalls.rs:196]: we have a validation on amount of declared syscalls there

File: utils/wasm-gen/src/wasm.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// WASM page has size of 64KiBs (65_536 bytes)
```
Comment (breathx) [inline utils/wasm-gen/src/wasm.rs:19]: doc comments all over the module ?

File: utils/wasm-gen/src/wasm.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+impl From<Page> for PageCount {
+    fn from(value: Page) -> Self {
+        Self(u32::from(value.0) + 1)
+    }
+}
+
+impl PageCount {
+    pub fn size(&self) -> u32 {
```
Comment (breathx) [inline utils/wasm-gen/src/wasm.rs:47]:
```suggestion
    pub fn mem_size(&self) -> usize {
```
`usize` usage is considering here

File: utils/wasm-gen/src/wasm.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        self.0 * PAGE_SIZE
+    }
+
+    pub fn raw_number(&self) -> u32 {
```
Comment (breathx) [inline utils/wasm-gen/src/wasm.rs:51]:
```suggestion
    pub fn raw(&self) -> u32 {
```

File: utils/wasm-gen/src/lib.rs [hunk 6]
```diff
@@ -236,25 +242,95 @@ pub fn gen_wasm_smith_module(u: &mut Unstructured, config: &SwarmConfig) -> Modu
...
+                    code.push(Instruction::I32Const((pointer_beyond - offset) as i32));
+                }
+            }
+
+            Parameter::MemoryValue => {
+                if unchecked {
+                    code.push(Instruction::I32Const(u.arbitrary().unwrap()));
+                } else {
+                    let memory_size = memory_pages.size();
+                    // u/i128 are the biggest values with 8 bytes size.
```
Comment (breathx) [inline utils/wasm-gen/src/lib.rs:288]: what about our repr(C) structs from gsys? e.g. LengthWithTwoHashes, that uses 32+32+4 = 68 bytes?

File: utils/wasm-gen/src/lib.rs [hunk 7]
```diff
@@ -236,25 +242,95 @@ pub fn gen_wasm_smith_module(u: &mut Unstructured, config: &SwarmConfig) -> Modu
...
 ) -> Vec<Instruction> {
-    let mut code = Vec::new();
-    for (index, val) in params.iter().enumerate() {
-        let rule = params_rules.get(index).cloned().unwrap_or_default();
+    let mut code = Vec::with_capacity(params.len() + 1 + results.len());
+    for val in params {
         let instr = match val {
-            ValueType::I32 => Instruction::I32Const(rule.get_i32(u)),
-            ValueType::I64 => Instruction::I64Const(rule.get_i64(u)),
+            ValueType::I32 => Instruction::I32Const(u.arbitrary().unwrap()),
```
Comment (breathx) [inline utils/wasm-gen/src/lib.rs:325]: expects, please

File: utils/wasm-gen/src/syscalls.rs [hunk 8]
```diff
@@ -136,7 +196,9 @@ pub(crate) fn sys_calls_table(config: &GearConfig) -> BTreeMap<SysCallName, SysC
         .map(|name| {
             (
                 name,
-                SysCallInfo::new(config, name.signature(), config.sys_call_freq),
+                SysCallInfo::new(config, name.signature(), config.sys_call_freq, {
+                    name == SysCallName::SendInput || name == SysCallName::SendInputWGas
```
Comment (breathx) [inline utils/wasm-gen/src/syscalls.rs:200]:
```suggestion
                    (SysCallName::SendInput, SysCallName::SendInputWGas, SysCallName::ReplyInput, SysCallName::ReplyInputWGas).contains(&name)
```

Comment (breathx) [reply utils/wasm-gen/src/syscalls.rs:200]: and also why?

**Review by breathx — APPROVED** [review]

File: utils/wasm-gen/src/lib.rs (hunk 7 above)
Comment (gshep) [reply utils/wasm-gen/src/lib.rs:325]: ok

File: utils/wasm-gen/src/lib.rs (hunk 6 above)
Comment (gshep) [reply utils/wasm-gen/src/lib.rs:288]: good catch, thanks! Will use some hard-coded value, 100 for example.

File: utils/wasm-instrument/src/syscalls.rs (hunk 2 above)
Comment (gshep) [reply utils/wasm-instrument/src/syscalls.rs:196]: The answer is that the method is used only in one test and in wasm-gen.

File: utils/wasm-gen/src/wasm.rs (hunk 4 above)
Comment (gshep) [reply utils/wasm-gen/src/wasm.rs:47]: unfortunately `usize` would force additional casts in memory pages checkings

**Pull request merged**